radixdlt/src/main/java/com/radixdlt/consensus/safety/SafetyRules.java

```diff
@@ -108,18 +113,17 @@ public ViewTimeout viewTimeout(View view, HighQC highQC) {
 	 * @param timestamp timestamp to use for the vote in milliseconds since epoch
 	 * @param highQC our current sync state
 	 * @return A vote result containing the vote and any committed vertices
-	 * @throws SafetyViolationException In case the vertex would violate a safety invariant
 	 */
-	public Vote voteFor(VerifiedVertex proposedVertex, BFTHeader proposedHeader, long timestamp, HighQC highQC) throws SafetyViolationException {
+	public Optional<Vote> voteFor(VerifiedVertex proposedVertex, BFTHeader proposedHeader, long timestamp, HighQC highQC) {
```

Comment:
Options can be quite cryptic. Future user of this method will have no clue what empty means. Is it a valid scenario? A failure?
We could add a bit more context here with Either (for example this: https://www.baeldung.com/vavr-either).
something like:
```
public Either<ConsensusError, Vote> voteFor(...) {
  if (...) {
    return Either.left(ConsensusError.VIOLATES_EARLIER_VOTE);
  }
  return Either.right(vote);
}
```
not just for this method but as a general approach.
wdyt?

Reply:
Yes, I think a good idea. Perhaps for another PR.